### PR TITLE
fix(dynamite)!: properly deserialize headers

### PR DESCRIPTION
**Pull request by Leptopoda**
Allows arbitrary objects in headers.
Also fixes list header deserialization

### CODE REVIEW

**Review by provokateurin — COMMENTED** [review]
I couldn't find anything wrong (other than the typos in the commit messages), but I'll give it another look when I'm more awake.

File: packages/dynamite/dynamite_runtime/lib/src/utils/codecs.dart [hunk 1]
```diff
@@ -0,0 +1,112 @@
+// ignore_for_file: public_member_api_docs
```
Comment (Leptopoda) [inline packages/dynamite/dynamite_runtime/lib/src/utils/codecs.dart:1]: Huh, totally forgot to remove this and document the interface before submitting.

File: packages/dynamite/dynamite_runtime/lib/src/utils/codecs.dart [hunk 2]
```diff
@@ -0,0 +1,132 @@
+import 'dart:convert';
+
+import 'package:built_collection/built_collection.dart';
+import 'package:built_value/serializer.dart';
+import 'package:dynamite_runtime/models.dart';
+
+/// A [HeaderCodec] encodes the json representation of a `Header` objects to
+/// strings and decodes strings to serialized `Header` objects.
```
Comment (provokateurin) [inline packages/dynamite/dynamite_runtime/lib/src/utils/codecs.dart:8]:
```suggestion
/// A [HeaderCodec] encodes the json representation of a [Header] objects to
/// strings and decodes strings to serialized [Header] objects.
```

**Review by provokateurin — CHANGES_REQUESTED** [review]

File: packages/dynamite/dynamite_runtime/lib/src/utils/codecs.dart [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        specifiedType: specifiedType,
+      );
+
+  @override
+  HeaderEncoder get encoder => HeaderEncoder(
+        explode: explode,
+      );
+}
+
+/// This class converts strings to serialized `Header` JSON objects.
```
[the remark this answers is not in the document]
Comment (Leptopoda) [reply packages/dynamite/dynamite_runtime/lib/src/utils/codecs.dart:43]: The Header type is not defined in the scope of this file.

File: packages/dynamite/dynamite_runtime/lib/src/utils/codecs.dart (hunk 2 above)
Comment (Leptopoda) [reply packages/dynamite/dynamite_runtime/lib/src/utils/codecs.dart:8]: The Header type is not defined in the scope of this file.

File: packages/dynamite/dynamite_runtime/lib/src/utils/codecs.dart [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        } else if (content.isEmpty) {
+          return null;
+        }
+
+        return jsonDecode(content);
+    }
+  }
+}
+
+/// This class converts serialized `Header` JSON objects to strings.
```
[the remark this answers is not in the document]
Comment (Leptopoda) [reply packages/dynamite/dynamite_runtime/lib/src/utils/codecs.dart:97]: The Header type is not defined in the scope of this file.

File: packages/dynamite/dynamite_runtime/lib/src/utils/codecs.dart (hunk 2 above)
Comment (provokateurin) [reply packages/dynamite/dynamite_runtime/lib/src/utils/codecs.dart:8]: It imports the models file so it should be?

Comment (Leptopoda) [reply packages/dynamite/dynamite_runtime/lib/src/utils/codecs.dart:8]: My bad sorry.
I didn't remember that I changed this last night :sweat_smile:

Comment (provokateurin) [reply packages/dynamite/dynamite_runtime/lib/src/utils/codecs.dart:8]: np

**Review by provokateurin — APPROVED** [review]

**Pull request merged**